### PR TITLE
Implement actor variations.

### LINKED ISSUE #15: Unit variations AKA Generals.
Had a draft during OpenRA-mods-should-soft-fork discussion. Might be hunting down all actorspawner cases to expand them with a PreventVariation tag is the most hardest point.

### PR DESCRIPTION
**Pull request by GraionDilach**
This should tackle the main reason the fork was started.

Fixes #15.

### CODE REVIEW

**Comment by GraionDilach** [comment]
At the moment, not all traits are expanded yet and I will need to look at Lua methods as well.

**Comment by GraionDilach** [comment]
NTM: Travis issue is due to code style.

File: OpenRA.Game/Actor.cs [hunk 1]
```diff
@@ -96,6 +96,16 @@ internal Actor(World world, string name, TypeDictionary initDict)
 					throw new NotImplementedException("No rules definition for unit " + name);
 
 				Info = world.Map.Rules.Actors[name];
+
+				if (Info.HasTraitInfo<VariationsInfo>())
```
Comment (abcdefg30) [inline OpenRA.Game/Actor.cs]: Upstream we'd usually use `TraitInfoOrDefault` instead, e.g. like
```
var variantInfo = Info.TraitInfoOrDefault<VariationsInfo>();
if (variantInfo != null)
{
	[...]
		name = variantInfo.Variations...;
	[...]
}
```

Don't know how you want to handle that in AS/Engine.

File: OpenRA.Game/Traits/Variations.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
```
Comment (abcdefg30) [inline OpenRA.Game/Traits/Variations.cs]: Not sure if that is true now. :P

File: OpenRA.Game/Traits/Variations.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ */
+#endregion
+
+namespace OpenRA.Traits
+{
+	[Desc("This actor has variations.")]
+	public class VariationsInfo : ITraitInfo, IRulesetLoaded
+	{
+		[FieldLoader.Require, Desc("Defines the list of variations this actor can turn into. Listing an entry multiple times will increase it's chance.")]
+		public string[] Variations = { };
```
Comment (abcdefg30) [inline OpenRA.Game/Traits/Variations.cs]: You could make this `readonly` I guess.

File: OpenRA.Game/Traits/Variations.cs [hunk 4]
```diff
@@ -12,12 +12,21 @@
...
 		public string[] Variations = { };
 
 		public object Create(ActorInitializer init) { return new Variations(init.Self, this); }
+
+		void IRulesetLoaded<ActorInfo>.RulesetLoaded(Ruleset rules, ActorInfo info)
+		{
+			foreach (var item in Variations)
+			{
+				if (!rules.Actors.ContainsKey(item.ToLowerInvariant()))
+					throw new YamlException("Invalid variation {0} defined!".F(item.ToLowerInvariant()));
```
Comment (GraionDilach) [inline OpenRA.Game/Traits/Variations.cs]: This should also have the actor's name in it as well.

**Comment by ABrandau** [comment]
If this is ever picked up, is there any chance it can also be implemented for buildings? So you could "rotate" a building by having it switch to actors from a list.

**Comment by GraionDilach** [comment]
NTM - look into the IceReaper system to see if that allows everything requires for this.

**Comment by GraionDilach** [comment]
Outdated and I probably won't resurrect this one in the short-to-middle term. `TransformOnCondition` + that random condition granter can be used as a workaround.

**Pull request closed without merging**